velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -135,79 +124,76 @@ struct RegexpReplaceFunction {
-      VELOX_USER_CHECK_GT(
-          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
-      ++numCodePoints;
-      utf8Position += charLength;
+    VELOX_USER_CHECK_GE(position, 1, "regexp_replace requires a position >= 1");
+    if (stringInput.size() == 0) {
+      if (pattern.size() == 0 && position == 1) {
+        result = replace;
+        return;
+      }
+      if (pattern.size() > 0) {
+        result = stringInput;
+        return;
+      }
     }
+
+    size_t utf8Position = getUTFLength(stringInput, position - 1);
     if (utf8Position > stringInput.size() + 1) {
-      result.resize(inputStringPiece.size());
-      std::memcpy(
-          result.data(), inputStringPiece.data(), inputStringPiece.size());
+      result = stringInput;
       return;
     }
 
-    re2::StringPiece prefix(inputStringPiece.data(), utf8Position);
-    re2::StringPiece targetStringPiece(
-        inputStringPiece.data() + utf8Position,
-        inputStringPiece.size() - utf8Position);
+    re2::RE2* patternRegex = getRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
 
+    std::string prefix(stringInput.data(), utf8Position);
     std::string targetString(
-        targetStringPiece.data(), targetStringPiece.size());
-    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+        stringInput.data() + utf8Position, stringInput.size() - utf8Position);
 
-    if (targetString.size() || prefix.size()) {
-      result.resize(prefix.size() + targetString.size());
-      std::memcpy(result.data(), prefix.data(), prefix.size());
-      std::memcpy(
-          result.data() + prefix.size(),
-          targetString.data(),
-          targetString.size());
-    } else {
-      result.resize(0);
-    }
+    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+    result = prefix + targetString;
   }
 
  private:
-  re2::RE2* getCachedRegex(const std::string& pattern) const {
-    auto it = patternCache_.find(pattern);
-    if (it != patternCache_.end()) {
+  re2::RE2* getRegex(const std::string& pattern) const {
+    auto it = cache_.find(pattern);
+    if (it != cache_.end()) {
       return it->second.get();
     }
     VELOX_USER_CHECK_LT(
-        patternCache_.size(),
+        cache_.size(),
         kMaxCompiledRegexes,
         "regexp_replace hit the maximum number of unique regexes: {}",
         kMaxCompiledRegexes);
     checkForCompatiblePattern(pattern, "regexp_replace");
     auto patternRegex = std::make_unique<re2::RE2>(pattern);
     auto* rawPatternRegex = patternRegex.get();
     checkForBadPattern(*rawPatternRegex);
-    patternCache_.emplace(pattern, std::move(patternRegex));
+    cache_.emplace(pattern, std::move(patternRegex));
     return rawPatternRegex;
   }
 
-  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>>
-      patternCache_;
+  template <typename TString>
+  size_t getUTFLength(TString& input, size_t maxCharacters) {
```

Comment:
It would be nice to move this function next to other UTF-8 functions, document it and add a dedicated unit test.

can input be a const &?

Reply:
Tested locally with const TString& input, everything compiled and all tests passed.

To clarify, do you mean moving the function next to other UTF-8 functions in velox/functions/lib/string/StringImpl.h?

I think that would be pretty straightforward and I can include it later tonight. Do you feel that would be better as its own PR since it is out of the scope of SparkSQL regexp_replace? I am happy with either solution.

In that context, would a better name for this function be "cappedUTFBytes/cappedUnicodeBytes"?

Reply:
I really love the way you go extra mile and really explore the space of possibilities. Thank you for doing this.

I'm thinking that this function could go next to cappedLength and perhaps, it can be named cappedByteLength to allow for distinguishing between length in characters vs. length in bytes. It can be part of this PR.